rust/opendp/src/trans/manipulation.rs

```diff
@@ -0,0 +1,231 @@
+            Function::new(|arg: &T| arg.clone()),
+            metric.clone(),
+            metric,
+            StabilityRelation::new_from_constant(Q::one())))
+    }
+}
+
+pub struct Clamp<M, T, Q> {
+    metric: PhantomData<M>,
+    data: PhantomData<T>,
+    distance: PhantomData<Q>
+}
+
+impl<M, T, Q> MakeTransformation2<VectorDomain<AllDomain<T>>, VectorDomain<IntervalDomain<T>>, M, M, T, T> for Clamp<M, Vec<T>, Q>
+    where M: Metric<Distance=Q>,
+          T: 'static + Clone + PartialOrd,
+          Q: 'static + One + Mul<Output=Q> + Div<Output=Q> + PartialOrd + DistanceCast {
+    fn make2(lower: T, upper: T) -> Fallible<Transformation<VectorDomain<AllDomain<T>>, VectorDomain<IntervalDomain<T>>, M, M>> {
+        Ok(Transformation::new(
+            VectorDomain::new_all(),
+            VectorDomain::new(IntervalDomain::new(Bound::Included(lower.clone()), Bound::Included(upper.clone()))),
+            Function::new(move |arg: &Vec<T>| arg.into_iter().map(|e| clamp(&lower, &upper, e)).collect()),
+            M::new(),
+            M::new(),
+            // clamping has a c-stability of one, as well as a lipschitz constant of one
+            StabilityRelation::new_from_constant(Q::one())))
+    }
+}
+
+impl<M, T, Q> MakeTransformation2<AllDomain<T>, IntervalDomain<T>, M, M, T, T> for Clamp<M, T, Q>
+    where M: Metric<Distance=Q>,
+          T: 'static + Clone + PartialOrd,
+          Q: 'static + One + Mul<Output=Q> + Div<Output=Q> + PartialOrd + DistanceCast {
+    fn make2(lower: T, upper: T) -> Fallible<Transformation<AllDomain<T>, IntervalDomain<T>, M, M>> {
+        Ok(Transformation::new(
+            AllDomain::new(),
+            IntervalDomain::new(Bound::Included(lower.clone()), Bound::Included(upper.clone())),
+            Function::new(move |arg: &T| clamp(&lower, &upper, arg)),
+            M::new(),
+            M::new(),
+            // clamping has a c-stability of one, as well as a lipschitz constant of one
+            StabilityRelation::new_from_constant(Q::one())))
+    }
+}
+
+fn clamp<T: Clone + PartialOrd>(lower: &T, upper: &T, x: &T) -> T {
+    (if x < &lower { lower } else if x > &upper { upper } else { x }).clone()
+}
+
+pub struct Unclamp<M, T, Q> {
+    metric: PhantomData<M>,
+    data: PhantomData<T>,
+    distance: PhantomData<Q>
+}
+
+impl<M, T, Q> MakeTransformation2<VectorDomain<IntervalDomain<T>>, VectorDomain<AllDomain<T>>, M, M, T, T> for Unclamp<M, Vec<T>, Q>
+    where M: Metric<Distance=Q>,
+          T: 'static + Clone + PartialOrd,
+          Q: 'static + Default + DistanceCast + One + Div<Output=Q> + Mul<Output=Q> + PartialOrd {
+    fn make2(lower: T, upper: T) -> Fallible<Transformation<VectorDomain<IntervalDomain<T>>, VectorDomain<AllDomain<T>>, M, M>> {
+        Ok(Transformation::new(
+            VectorDomain::new(IntervalDomain::new(Bound::Included(lower), Bound::Included(upper))),
+            VectorDomain::new_all(),
+            Function::new(move |arg: &Vec<T>| arg.clone()),
```

Comment:
Seeing these clones here makes me more convinced that we should move to slices for input and output. (Later!)